File: ruby/17-taxonomy-trap/src/vehicle.rb
# frozen_string_literal: true

class Vehicle
  attr_reader :brand, :model, :year

  def initialize(brand, model, year)
    @brand = brand
    @model = model
    @year = year
  end
end

class Car < Vehicle
  def daily_rate
    40
  end

  def insurance_cost(days)
    days * 12
  end
end

class Truck < Vehicle
  def daily_rate
    80
  end

  def insurance_cost(days)
    days * 20
  end
end

class ElectricCar < Car
  def fuel_cost(_days)
    0
  end

  def rental_total(days, gps = false)
    (daily_rate * days) + fuel_cost(days) + insurance_cost(days) + (gps ? 8 * days : 0)
  end
end

class DieselCar < Car
  def fuel_cost(days)
    days * 5
  end

  def rental_total(days, gps = false)
    (daily_rate * days) + fuel_cost(days) + insurance_cost(days) + (gps ? 8 * days : 0)
  end
end

class ElectricTruck < Truck
  def fuel_cost(_days)
    0
  end

  def rental_total(days, gps = false)
    (daily_rate * days) + fuel_cost(days) + insurance_cost(days) + (gps ? 8 * days : 0)
  end
end

class DieselTruck < Truck
  def fuel_cost(days)
    days * 15
  end

  def rental_total(days, gps = false)
    (daily_rate * days) + fuel_cost(days) + insurance_cost(days) + (gps ? 8 * days : 0)
  end
end

File: javascript/17-taxonomy-trap/src/vehicle.js
class Vehicle {
  constructor(brand, model, year) {
    this.brand = brand;
    this.model = model;
    this.year = year;
  }
}

class Car extends Vehicle {
  daily_rate() {
    return 40;
  }

  insurance_cost(days) {
    return days * 12;
  }
}

class Truck extends Vehicle {
  daily_rate() {
    return 80;
  }

  insurance_cost(days) {
    return days * 20;
  }
}

class ElectricCar extends Car {
  fuel_cost(days) {
    return 0;
  }

  rental_total(days, gps = false) {
    return this.daily_rate() * days + this.fuel_cost(days) + this.insurance_cost(days) + (gps ? 8 * days : 0);
  }
}

class DieselCar extends Car {
  fuel_cost(days) {
    return days * 5;
  }

  rental_total(days, gps = false) {
    return this.daily_rate() * days + this.fuel_cost(days) + this.insurance_cost(days) + (gps ? 8 * days : 0);
  }
}

class ElectricTruck extends Truck {
  fuel_cost(days) {
    return 0;
  }

  rental_total(days, gps = false) {
    return this.daily_rate() * days + this.fuel_cost(days) + this.insurance_cost(days) + (gps ? 8 * days : 0);
  }
}

class DieselTruck extends Truck {
  fuel_cost(days) {
    return days * 15;
  }

  rental_total(days, gps = false) {
    return this.daily_rate() * days + this.fuel_cost(days) + this.insurance_cost(days) + (gps ? 8 * days : 0);
  }
}

export { Vehicle, Car, Truck, ElectricCar, DieselCar, ElectricTruck, DieselTruck };
